HPCC-13253 UNICODE EditDistance Enhancements

emuharemagic opened the pull request:
- Modify ulUnicodeLocaleEditDistance() to use the most recent version of code V4 instead of the V2
- Add support for fast, non-locale versions of EditDistance and EditDistanceWithinRadius functions; The non-locale implementations will be based on an assumption that every user-perceived Unicode character can be encoded with no more than 2 16-bit words;
- Update TestEditDistance.ecl and TestEditDistanceWithinRadius.ecl to test new editDistance code

Signed-off-by: Edin Muharemagic edin.muharemagic@lexisnexis.com

emuharemagic commented:
@ghalliday - please review

AttilaVamos commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -593,9 +614,10 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
-    unsigned minED = (leftLen < rightLen)? rightLen - leftLen: leftLen - rightLen;
-    if (minED > radius)
-        return minED;
+    // this shortcut is not applicable any longer because unicode characters could take 2 or more UChars
```
Is this shortcut valid if bi = NULL?

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -286,28 +292,43 @@ class UPCList // User perceived character list
         : length_(0), capacity_(capacity),ustring_(source), invalid_(false)
     {
-        doCreateUPCList(cbi);
+        (cbi == (const BreakIterator*)0)?doCreateUPCList():doCreateUPCList(cbi);
```
Trivial: More normal to write as

!cbi?doCreateUPCList():doCreateUPCList(cbi);

ghalliday replied:
or

  if (cbi)
    doCreateUPCList()
  else;
    doCreateUPCList(cbi);

Since they are functions that return void.

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -286,28 +292,43 @@ class UPCList // User perceived character list
+        {
+            end = end+ucpLength(ustring_[end]);
+            next_[++index] = end;
+            length_++;
```
trivial: Could have length_ = index; after the loop has finished.

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -286,28 +292,43 @@ class UPCList // User perceived character list
     uint32_t  capacity_;
     bool invalid_;
-    void doCreateUPCList(BreakIterator& cbi) {
+    void doCreateUPCList(BreakIterator* cbi) {
```
minor: Better to keep this as a reference parameter and pass *cnbi in.  (It clarifies this function does not support NULL).

ghalliday commented:
@emuharemagic looks good, and I would be happy to merge as is.  Let me know if you want to change or merge as-is.

emuharemagic commented:
Thank you for your comments - I will make suggested changes

Commit:
HPCC-13253 UNICODE EditDistance Enhancements

- Modify ulUnicodeLocaleEditDistance() to use the most recent version of code V4 instead of the V2
- Add support for fast, non-locale versions of EditDistance and
EditDistanceWithinRadius functions; The non-locale implementations will be based on an assumption that every user-perceived Unicode character can be
encoded with no more than 2 16-bit words;
- Update TestEditDistance.ecl and TestEditDistanceWithinRadius.ecl to test new editDistance code


Signed-off-by: Edin Muharemagic <edin.muharemagic@lexisnexis.com>

emuharemagic commented:
@ghalliday - I made the changes - pls review

AttilaVamos commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK
